rpki-validator/src/main/java/net/ripe/rpki/validator3/rrdp/RrdpServiceImpl.java

```diff
@@ -116,23 +124,16 @@ private boolean doStoreRepository(RpkiRepository rpkiRepository, RpkiRepositoryV
         if (notification.sessionId.equals(rpkiRepository.getRrdpSessionId())) {
             if (rpkiRepository.getRrdpSerial().compareTo(notification.serial) <= 0) {
                 try {
-                    final List<Delta> deltas = notification.deltas.stream().
-                            filter(d -> d.getSerial().compareTo(rpkiRepository.getRrdpSerial()) > 0).
-                            sorted(Comparator.comparing(DeltaInfo::getSerial)).
-                            parallel().
-                            map(di -> readDelta(notification, di)).
-                            collect(Collectors.toList());
-
-                    verifyDeltaSerials(deltas, notification, rpkiRepository);
-
-                    deltas.forEach(d -> {
-                        storage.readTx0(tx -> verifyDeltaIsApplicable(tx, d));
-                        storage.writeTx0(tx -> {
-                            storeDelta(tx, d, validationRun, rpkiRepository, changedObjects);
-                            tx.afterCommit(() -> rpkiRepository.setRrdpSerial(rpkiRepository.getRrdpSerial().add(BigInteger.ONE)));
-                        });
-                    });
-
+                    List<DeltaInfo> orderedDeltas = verifyDeltaSerials(notification, rpkiRepository);
+                    rrdpProcessingPool.submit(() -> orderedDeltas
+                            .parallelStream()
+                            .map(di -> readDelta(notification, di))
+                            .forEachOrdered(d -> storage.writeTx0(tx -> {
+                                verifyDeltaIsApplicable(tx, d);
```

Comment:
Since you changing this code, is it possible to move all the writeTx(...) code into a separate pool with one thread? All writing transactions are exclusive, so they block the database and all this parallelism is not so effective.

What we probably want is "4 threads are downloading and parsing, putting results in the queue and 1 thread reading the queue and saving to the database". The "queue" doesn't have to be explicit, can be part of a pool, of course.

Unless it's too complicated.

Reply:
Ah, it's probably already done by forEachOrdered! Disregard my question then.